src/qr/dclogin_scheme.rs

```diff
@@ -0,0 +1,402 @@
+use std::collections::HashMap;
+
+use crate::config::Config;
+use crate::context::Context;
+use crate::provider::Socket;
+use crate::{contact, login_param::CertificateChecks};
+use anyhow::{bail, Context as _, Result};
+use num_traits::cast::ToPrimitive;
+
+use super::{Qr, DCLOGIN_SCHEME};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LoginOptions {
+    UnsuportedVersion(u32),
+    V1 {
+        mail_pw: String,
+        imap_host: Option<String>,
+        imap_port: Option<u16>,
+        imap_username: Option<String>,
+        imap_password: Option<String>,
+        imap_security: Option<Socket>,
+        imap_certificate_checks: Option<CertificateChecks>,
+        smtp_host: Option<String>,
+        smtp_port: Option<u16>,
+        smtp_username: Option<String>,
+        smtp_password: Option<String>,
+        smtp_security: Option<Socket>,
+        smtp_certificate_checks: Option<CertificateChecks>,
+    },
+}
+
+/// scheme: `dclogin://user@host/?p=password&v=1[&options]`
+/// read more about the scheme at https://github.com/deltachat/interface/blob/master/uri-schemes.md#DCLOGIN
+pub(super) fn decode_login(qr: &str) -> Result<Qr> {
+    let url = url::Url::parse(qr).with_context(|| format!("Malformed url: {:?}", qr))?;
+
+    let mut payload = qr
+        .get(DCLOGIN_SCHEME.len()..)
+        .context("invalid DCLOGIN payload E1")?;
+
+    // if first 2 chars are `//` remove them
+    if payload.get(0..2) == Some("//") {
+        payload = payload.get(2..).context("invalid DCLOGIN payload E2")?;
+        // todo: is there a more idiomatic way?
+    }
+
+    let addr = payload
+        .get(
+            ..payload
+                .chars()
+                .position(|c| c == '?' || c == '/')
+                .context("invalid DCLOGIN payload E3a")?,
+        )
+        .context("invalid DCLOGIN payload E3b")?;
+
+    let mut scheme = url.scheme().to_owned();
+    scheme.make_ascii_lowercase();
+
+    if scheme == "dclogin" {
+        let options = url.query_pairs();
+        if options.count() == 0 {
+            bail!("invalid DCLOGIN payload E4")
+        }
+        // load options into hashmap
+        let mut parameter_map = HashMap::with_capacity(options.count());
+        for (key, value) in options {
+            parameter_map.insert(key.into_owned(), value.into_owned());
+        }
+
+        // check if username is there
+        println!("{}", addr);
+        if !contact::may_be_valid_addr(addr) {
+            bail!("invalid DCLOGIN payload: invalid username E5");
+        }
+
+        // apply to result struct
+        let options: LoginOptions = match parameter_map.get("v").map(|i| i.parse::<u32>()) {
+            Some(version_result) => match version_result {
+                Ok(1) => LoginOptions::V1 {
```

Comment:
`match`es don't need to be nested, you can do `Some(Ok(1)) => {...}, Some(Ok(v)) => LoginOptions::UnsuportedVersion(v),` and so on